### PR TITLE
KAFKA-8700: Flaky Test QueryableStateIntegrationTest#queryOnRebalance

### PR DESCRIPTION
**Pull request by cpettitt-confluent**
This is not guaranteed to actually fix `queryOnRebalance`, since the
failure could never be reproduced locally. I did not bump timeouts
because it looks like that has been done in the past for this test
without success. Instead this change makes the following improvements:

1. It waits for the application to be in a RUNNING state before
proceeding with the test.

2. It waits for the remaining instance to return to RUNNING state
within a timeout after rebalance. I observed once that we were able to
do the KV queries but the instance was still in REBALANCING, so this
should reduce some opportunity for flakiness.

3. The meat of this change: we now iterate over all keys in one shot
(vs. one at a time with a timeout) and collect various failures, all of
which are reported at the end. This should help us to narrow down the
cause of flakiness if it shows up again.

Before:

```
java.lang.AssertionError: Condition not met within timeout 120000. waiting for metadata, store and value to be non null
    at org.apache.kafka.test.TestUtils.waitForCondition(TestUtils.java:376)
    ...
```

After:

```
java.lang.AssertionError: Not all keys are available for store word-count-store-stream-three-0 in 120000 ms
    * No value is available for these keys: [goodbye, foo, bar]
    * Exceptions were raised for the following keys:
        Sam:
            java.lang.IllegalArgumentException: Ooops
            	at org.apache.kafka.streams.integration.QueryableStateIntegrationTest.lambda$verifyAllKVKeys$6(QueryableStateIntegrationTest.java:289)
            	at org.apache.kafka.test.TestUtils.retryOnExceptionWithTimeout(TestUtils.java:417)
            	at org.apache.kafka.test.TestUtils.retryOnExceptionWithTimeout(TestUtils.java:385)
            	at org.apache.kafka.streams.integration.QueryableStateIntegrationTest.verifyAllKVKeys(QueryableStateIntegrationTest.java:265)
            	at org.apache.kafka.streams.integration.QueryableStateIntegrationTest.queryOnRebalance(QueryableStateIntegrationTest.java:431)

	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:24)
        ...
```

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/test/java/org/apache/kafka/test/TestUtils.java
Comment (bbejeck) [inline clients/src/test/java/org/apache/kafka/test/TestUtils.java:419]: nit: the method is named `retryOnExceptionWithTimeout` but it will always throw with the `NoRetryException`.  Should the name of the method change to something like `maybeRetryOnExceptionWithTimeout

**Comment by cpettitt-confluent** [comment]
11/2.13 Green.

11/2.21: kafka.api.AdminClientIntegrationTest.testElectUncleanLeadersAndNoop

8/2.11: kafka.api.PlaintextConsumerTest.testLowMaxFetchSizeForRequestAndPartition

This PR is test only and none of the added tests were in the failure list.

**Review by bbejeck — APPROVED** [review]
Thanks for this @cpettitt-confluent I just have one minor comment, otherwise LGTM

**Comment by bbejeck** [comment]
call for second review one of @mjsax @abbccdda

**Comment by cpettitt-confluent** [comment]
Thanks for the review @bbejeck!

**Comment by guozhangwang** [comment]
cc @bbejeck @ableegoldman @abbccdda

File: streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java]: Why we wait for state after the verification than before it?

File: streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java:259]: Good call!

**Review by guozhangwang — COMMENTED** [review]
Just minor question, otherwise lgtm.

File: streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java
Comment (cpettitt-confluent) [reply streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java]: Only because that is the way it was in the original test. It also turned up the interesting fact that the stores are queryable before the instance is in a RUNNING state. It seems OK to be in either order provided that we're giving enough time to get to a running state, but I can re-arrange them if that is needed to get this committed.

Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java]: Since we merged in KIP-429, under normal processing the state of streams would be in REBALANCING for very short time in between poll calls only during restoration, I think that's why you observed that it is queryable, as the state transition to RUNNING happens instantaneously after line 454 (i.e. the restoration takes no time).

Just for the logical reasoning itself, I think it is still better to check state transition first, and then verify queryable results.

**Comment by bbejeck** [comment]
Java 8 and Java 11/2.12 failed test results already cleaned up.

Java 11/2.13 passed

retest this please

**Comment by bbejeck** [comment]
retest this please

**Comment by cpettitt-confluent** [comment]
I've switched the order of the assertions per your request @guozhangwang. Can you please re-review?

**Comment by guozhangwang** [comment]
LGTM! Thanks @cpettitt-confluent , I think we can merge it now.

**Pull request merged**
